internal/telemetry/appinsights.go

```diff
@@ -0,0 +1,67 @@
+package telemetry
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/microsoft/ApplicationInsights-Go/appinsights"
+)
+
+var telemetryClient appinsights.TelemetryClient
+
+func SetTelemetryClient(client appinsights.TelemetryClient) {
+	telemetryClient = client
+}
+
+func SetTelemetryClientFromInstrumentationKey(instrumentationKey string) {
+	telemetryClient = appinsights.NewTelemetryClient(instrumentationKey)
+}
+
+type Telemetry struct {
+	Client appinsights.TelemetryClient
+}
+
+func InitializeAppInsights() {
+	instrumentationKey := "3fdeec77-2951-456f-bcb2-9f003b512a3f"
+	config := appinsights.NewTelemetryConfiguration(instrumentationKey)
+	telemetryClient = appinsights.NewTelemetryClientFromConfig(config)
+	SetTelemetryClient(telemetryClient)
+	// Add a diagnostics listener for printing telemetry messages
+	appinsights.NewDiagnosticsMessageListener(func(msg string) error {
+		fmt.Printf("[%s] %s\n", time.Now().Format(time.UnixDate), msg)
+		return nil
+	})
+}
+
+// TrackCommand tracks a command execution event
+func TrackCommand(command string) {
+	event := appinsights.NewEventTelemetry("command")
+	event.Properties["command"] = command
+	telemetryClient.Track(event)
+}
+
+// TrackSubCommand tracks a sub-command execution event
+func TrackSubCommand(command, subCommand string) {
+	event := appinsights.NewEventTelemetry("sub-command")
+	event.Properties["command"] = command
+	event.Properties["sub-command"] = subCommand
+	telemetryClient.Track(event)
+}
+
+func TrackEvent(eventName string) {
+	event := appinsights.NewEventTelemetry(eventName)
+	event.Properties["command"] = eventName
```

Comment:
>.Properties [](http://example.com/codeflow?start=6&length=11)

what kind of map is `Properties`? Can a value be an arbitrary Go object that implements a particular interface for serialization, or does it just take string values?

Reply:
This is of type string only. This is map of string keys and string values. We need to convert Go object as a value to the properties map by converting into string only .